### PR TITLE
Avoid eagerly failing connections when request closes

### PR DESCRIPTION
**Pull request by raboof**
Refs #2066

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala [hunk 1]
```diff
@@ -216,6 +221,11 @@ private[http] final class GracefulTerminatorStage(settings: ServerSettings)
           pendingUserHandlerResponse = true
           push(toUser, request)
         }
+
+        override def onUpstreamFinish(): Unit = {
+          // don't finish the whole bidi stage, just propagate the completion:
+          complete(toUser)
+        }
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala]: Do we have to keep track and stop the stage when both flows have completed?

Comment (johanandren) [reply akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala]: That was my first thought as well, if we keep the bidistage alive here, then when _does_ it actually stop

Comment (johanandren) [reply akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala]: Maybe do some debug printing in the stop callback and verify that something else makes it stop?

Comment (raboof) [reply akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala]: `postStop` seems to be called fine. Also https://doc.akka.io/docs/akka/current/stream/stream-customize.html#custom-processing-with-graphstage says "`completeStage()` is equivalent to closing all output ports and cancelling all input ports.", it sounds like this should be sufficient.

**Comment by raboof** [comment]
Verified with a load test (also had to merge in #2070)

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala (hunk 1 above)
Comment (johanandren) [reply akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala]: Thanks for verifying!

**Comment by ktoso** [comment]
PLS BUILD

**Comment by ktoso** [comment]
Failure was https://github.com/akka/akka-http/issues/2076 seems unrelated... running again

**Pull request merged**